[kalarisng] iP

kalarisng opened the pull request:
# DUKE
> "Everything begins with an idea." - Earl Nightingale

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
* easy to learn
+ ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

kalarisng committed:
Level-1

kalarisng committed:
Level-2

kalarisng committed:
Level-3

kalarisng committed:
Level-4

kalarisng committed:
A-TextUiTesting

kalarisng committed:
Level-5

kalarisng committed:
Level-6

kalarisng committed:
Level-8

kalarisng committed:
Level-7

kalarisng committed:
Merge Level-8

kalarisng committed:
Merge remote-tracking branch 'origin/add-gradle-support'
  Adds Gradle support to project

kalarisng committed:
Adds Gradle Support

kalarisng committed:
Adds more OOP

kalarisng committed:
Organise classes into packages

kalarisng committed:
Add JUnit Tests

kalarisng committed:
Add JavaDoc Comments

kalarisng committed:
Tweak code to follow coding standard

kalarisng committed:
Find tasks by searching for keyword

kalarisng committed:
Merge branch 'branch-A-CodingStandard'
Merges branch A-CodingStandard

kalarisng committed:
Merges Level-9

gitsac commented on src/main/java/duke/EmptyDescription.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Represents an Exception due to absence of description following task.
+ */
+public class EmptyDescription extends Exception {
```
I like that you abstracted empty user inputs into its own class!

gitsac commented on src/main/java/duke/WrongTask.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represents an Exception due to user inputting a wrong task keyword.
+ */
+public class WrongTask extends Exception {
```
Would it be better to centralize a DukeExceptions class which can then be inherited by the various separate Exception classes?

gitsac commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+public class Deadline extends Task {
+    protected final String by;
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+    }
+
+    /**
+     * Creates a new Deadline task.
+     * @param description Description of the task.
+     * @return Deadline task.
+     * @throws DateTimeParseException If the date is in the wrong format.
+     */
+    public static Deadline createDeadline(String description) throws DateTimeParseException {
+        String[] split = description.split("/by ");
+        LocalDate d1 = LocalDate.parse(split[1]);
+        return new Deadline(split[0], d1.format(DateTimeFormatter.ofPattern("MMM d yyyy")));
+    }
```
Would it be better to make the Deadline constructor private if there is going to be a factory method to create Deadline tasks?

gitsac reviewed (approved):
I like that you have abstracted the various exceptions for different cases into their own classes, very neat, but you might want to consider whether a centralized class for Exceptions pertaining to Duke whereby the other classes inherit from would bring it up a notch.
Overall, the rest of your iP looks good and well-documented in terms of your JavaDocs!

wpx12011 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,34 @@
+    private final TaskList taskList;
+    private final Ui ui;
+    private final Storage storage;
+    private static final String filePath = "data/duke.txt";
```
Should the constant filePath be named as 'FILEPATH'?

wpx12011 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+package duke;
```
Good job done for putting every class in a package!

wpx12011 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,190 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
```
Good job done for having specific import statement.

wpx12011 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,190 @@
+     * @return True or false depending on whether description of task is empty.
+     * @throws EmptyDescription If the description of task is empty.
+     */
+    public static boolean checkEmptyDescription(String[] checkString) throws EmptyDescription {
```
Perhaps the boolean method could be named as 'isEmpty'?
I think the method will always return true, and the usage of it is to throw an exception when the string description is empty. May be can set the method to be void so there won't be a need to return anything?

wpx12011 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Unmark the task as done.
+     */
+    public void unMarkAsDone() {
```
I could understand that the naming unMarkAsDone somewhat gives it a better contrast with markAsDone haha, but perhaps need to change to unmarkAsDone() since unmark is a word itself.

wpx12011 reviewed (approved):
Nice OOP code and I could follow the structures clearly, great job!

kalarisng committed:
Add GUI

kalarisng committed:
Improve Code Quality

There are several long methods in some classes,
making it unreadable.

Shortening the code by adding more short methods
will improve code quality.

Using more short methods is preferred over using
less long methods.

kalarisng committed:
Use assert feature

Document important information that should hold at various
points of the code.

Let's insert some assert features into the code.

Using assert helps with debugging as an assertion failure
indicates a possible bug in the code.

kalarisng committed:
Merge pull request #2 from kalarisng/branch-A-Assertions

Use assert feature

kalarisng committed:
Merge branch 'master' into branch-A-CodeQuality

kalarisng committed:
Merge pull request #3 from kalarisng/branch-A-CodeQuality

Improve Code Quality

kalarisng committed:
Add an extension - C-Update

Allow user to easily edit information.

Being able to edit information is more user-friendly
as user does not have to delete and re-add in the case
of inputting wrong information.

Using a method is preferable as we can simply
modify the method in the event of bug-fixing.

kalarisng committed:
Update JavaDocs

kalarisng committed:
Add Duke Greeting to GUI

kalarisng committed:
Add update ability for Deadline tasks

kalarisng committed:
Add More Error Handling

kalarisng committed:
Update README.md

kalarisng committed:
Update JavaDocs

kalarisng committed:
Merge branch 'master' of https://github.com/kalarisng/ip

kalarisng committed:
Update README.md

kalarisng committed:
Update README.md

kalarisng committed:
Update README.md

kalarisng committed:
Improve GUI

kalarisng committed:
Update UI personality

kalarisng committed:
Add title to window

kalarisng committed:
Update Ui.png

kalarisng committed:
Update exit method and create user guide

kalarisng committed:
Update user guide

kalarisng committed:
Update JavaDocs and Java coding standard

kalarisng committed:
Update Ui.png

kalarisng committed:
Update user guide and Java coding standards
